src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,708 @@
+            };
+
+            // Serialize
+            string serialized = JsonSerializer.Serialize(obj, s_optionReadAndWriteFromStr);
+
+            // Deserialize
+            obj = JsonSerializer.Deserialize<Class_With_NullableUInt64_And_Float>(serialized, s_optionReadAndWriteFromStr);
+
+            // Test roundtrip
+            JsonTestHelper.AssertJsonEqual(serialized, JsonSerializer.Serialize(obj, s_optionReadAndWriteFromStr));
+        }
+
+        private class Class_With_NullableUInt64_And_Float
+        {
+            public ulong? NullableUInt64Number { get; set; }
+            [JsonInclude]
+            public List<float> FloatNumbers;
+        }
+
+        [Fact]
+        public static void FloatingPointConstants()
+        {
+            // Valid values
+            PerformFloatingPointSerialization("NaN", quoted: true, "NaN");
+            PerformFloatingPointSerialization("Infinity", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("-Infinity", quoted: true, "-Infinity");
+
+            // Variations that Utf8Parser permits.
+            PerformFloatingPointSerialization("naN", quoted: true, "NaN");
+            PerformFloatingPointSerialization("Nan", quoted: true, "NaN");
+            PerformFloatingPointSerialization("NAN", quoted: true, "NaN");
+
+            PerformFloatingPointSerialization("+Infinity", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("+infinity", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("infinity", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("infinitY", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("INFINITY", quoted: true, "Infinity");
+            PerformFloatingPointSerialization("+INFINITY", quoted: true, "Infinity");
+
+            PerformFloatingPointSerialization("-infinity", quoted: true, "-Infinity");
+            PerformFloatingPointSerialization("-infinitY", quoted: true, "-Infinity");
+            PerformFloatingPointSerialization("-INFINITY", quoted: true, "-Infinity");
+
+            // Invalid values
+            PerformFloatingPointSerialization(" NaN");
+            PerformFloatingPointSerialization(" Infinity");
+            PerformFloatingPointSerialization(" -Infinity");
+            PerformFloatingPointSerialization("NaN ");
+            PerformFloatingPointSerialization("Infinity ");
+            PerformFloatingPointSerialization("-Infinity ");
+            PerformFloatingPointSerialization("a-Infinity");
+            PerformFloatingPointSerialization("NaNa");
+            PerformFloatingPointSerialization("Infinitya");
+            PerformFloatingPointSerialization("-Infinitya");
+        }
+
+        private static void PerformFloatingPointSerialization(string testString, bool quoted = true, string constantAsString = null)
+        {
+            string expectedJson;
+            string testJson;
+
+            if (quoted)
```

Comment:
Mentioned elsewhere as well: having two tests instead of this `if` statement would make it easier to understand IMO.